### PR TITLE
outdated --mode=null-safety, resolvable constrained to null-safe vers…

### PR DESCRIPTION
**Pull request by sigurdm**
…ions when they exist

### CODE REVIEW

File: lib/src/pubspec_utils.dart [hunk 1]
```diff
@@ -37,6 +38,62 @@ Pubspec stripDependencyOverrides(Pubspec original) {
   );
 }
 
+Future<Pubspec> constrainedToAtLeastNullSafetyPubspec(
+    Pubspec original, SystemCache cache) async {
+  /// Get the first version of [package] opting in to null-safety.
+  Future<VersionRange> _constrainToFirstWithNullSafety(
+      PackageRange packageRange) async {
+    final ref = packageRange.toRef();
```
Comment (jonasfj) [inline lib/src/pubspec_utils.dart:46]: Maybe return sources other than `HostedSource` without any changes.

File: lib/src/pubspec_utils.dart [hunk 2]
```diff
@@ -37,6 +38,62 @@ Pubspec stripDependencyOverrides(Pubspec original) {
...
+    final available = await cache.source(ref.source).getVersions(ref);
+    if (available.isEmpty) {
+      return packageRange.constraint;
```
Comment (jonasfj) [inline lib/src/pubspec_utils.dart:49]: Don't this need to `any`?

File: lib/src/pubspec_utils.dart [hunk 3]
```diff
@@ -37,6 +38,62 @@ Pubspec stripDependencyOverrides(Pubspec original) {
...
+
+    available.sort((x, y) => x.version.compareTo(y.version));
+
+    for (final p in available) {
+      final pubspec = await cache.source(ref.source).describe(p);
+      if (pubspec.languageVersion.supportsNullSafety) {
+        return VersionRange(min: p.version, includeMin: true);
+      }
+    }
+    return packageRange.constraint;
```
Comment (jonasfj) [inline lib/src/pubspec_utils.dart:60]: Don't this need to `any`?

File: lib/src/pubspec_utils.dart [hunk 4]
```diff
@@ -37,6 +38,62 @@ Pubspec stripDependencyOverrides(Pubspec original) {
...
+  }
+
+  Future<List<PackageRange>> atLeastNullsafety(
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/pubspec_utils.dart:63]: Yes, this applies the helper above to a list of packages (called once for each of deps and dev_deps)

File: lib/src/pubspec_utils.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/pubspec_utils.dart:46]: Those don't even get here.

File: lib/src/pubspec_utils.dart (hunk 3 above)
Comment (sigurdm) [reply lib/src/pubspec_utils.dart]: It should probably be at least current.

Done

File: lib/src/pubspec_utils.dart (hunk 2 above)
Comment (sigurdm) [reply lib/src/pubspec_utils.dart]: done

**Comment by sigurdm** [comment]
CC @mit-mit This implements what we discussed earlier today (it doesn't remove the upgradable column).

File: lib/src/pubspec_utils.dart [hunk 5]
```diff
@@ -37,6 +38,63 @@ Pubspec stripDependencyOverrides(Pubspec original) {
...
+      /// We only need to remove the upper bound if it is a hosted package.
+      if (packageRange.source is HostedSource) {
+        unconstrainedRange = PackageRange(
+            packageRange.name,
+            packageRange.source,
+            await constrainToFirstWithNullSafety(packageRange),
+            packageRange.description,
+            features: packageRange.features);
+      }
+      result.add(unconstrainedRange);
```
Comment (jonasfj) [inline lib/src/pubspec_utils.dart:81]: Consider making this parallel, otherwise it might be somewhat slow..

Comment (sigurdm) [reply lib/src/pubspec_utils.dart]: Good point
Done

**Review by jonasfj — APPROVED** [review]

**Comment by mit-mit** [comment]
I tried to run this on https://github.com/flutter/gallery. That produced:
https://gist.github.com/mit-mit/60943b4f8a6a94f68cdb176e3d07ea6f#file-new-outdated-txt
which has no null safety resolution (Resolvable is empty).

Then I tried to manually use the algorithm I think we discussed: For a dep with no null safety version, use `any`, and for those that had, use `^<version with nullsafety>`. That pubspec is here, and that does seem to resolve?

```
environment:
  sdk: ">=2.6.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter
  intl: ^0.17.0-nullsafety
  flutter_localized_locales: ^2.0.0-nullsafety
  cupertino_icons: ^1.0.0
  rally_assets:
  meta: ^1.3.0-nullsafety
  scoped_model:
  shrine_images:
  flare_dart:
  flare_flutter:
  url_launcher: ^6.0.0-nullsafety
  vector_math: ^2.1.0-nullsafety
  shared_preferences:
  collection: ^1.15.0-nullsafety
  flutter_gallery_assets:
  package_info:
  google_fonts:
  flutter_staggered_grid_view:
  animations:
  provider:
  adaptive_breakpoints:

dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_driver:
    sdk: flutter
  test: ^1.16.0-nullsafety
  path: ^1.8.0-nullsafety
  args:
  grinder:
  pedantic: ^1.10.0-nullsafety
  string_scanner: ^1.1.0-nullsafety
  web_benchmarks:
```

**Pull request merged**